### PR TITLE
INFRA-541 Build Orange Docker container

### PR DESCRIPTION
**Pull request by nedleitch**
Add the creation of a Docker container to the building of `hmftools` modules. Does not do the Docker part if there is no Dockerfile for the module.

I've leveraged the Python script to do the heavy lifting but since the containers used in the build environment are quite modular I opted to generate the Docker-related commands for execution by a subsequent stage. I feel given the simplicity of the build overall this is acceptable.

### CODE REVIEW

File: hmftools-build.py
Comment (mvanniekerkHartwig) [inline hmftools-build.py:51]: So if I understand correctly, in your approach you're choosing to release to an open-to-the-world artifact registry? Did you also consider releasing to docker central? (I think this approach is fine, but it could be good to also consider releasing to docker central for visibility.)

**Review by mvanniekerkHartwig — COMMENTED** [review]
Code looks good, I think it will work just fine! I have 2 concerns though, that you probably thought about as well, so I'm curious to hear your thoughts:

1. To make this project succeed, we should also include tools (and tool versions) that were released in the past (maybe a long time ago). How are you planning to deal with those, since you are working from the master branch of hmftools? With this approach it will be tough to "go back in history".
2. Not everyone that releases tools in this repo will stick to "our" tag-release style. Also, there is no guarantee that everyone that releases tools on this repo will include the _same_ jar in the [releases page](https://github.com/hartwigmedical/hmftools/releases) that you get from running `mvn clean install` on the commit that was tagged. I believe there have been (please correct me if I'm wrong) multiple cases in the past where the commit that was tagged with the release does not compile or does not pass tests, and the jar that was included in the releases page was actually generated from a different commit, or from a code state that was never committed. Therefore to work this approach may require some behaviour changes that may or may not be feasible to ask from all contributors on this project.

Because of points 1 and 2, I advocated an approach of putting the docker files in a separate repo, and adding some "subscription" logic, that creates a new docker image every time a new jar is added to the [releases page](https://github.com/hartwigmedical/hmftools/releases). Did you also consider this approach? I'd like to hear your thoughts on this, how you are going to address point 1, and whether you think point 2 is a real problem or can be ignored.

File: hmftools-build.py
Comment (nedleitch) [reply hmftools-build.py:53]: Haven't quite pinned down the external repo, I'd be just fine with Docker central as well, but the principal requirement here was that Stephen (Watts, in Melbourne, who has built the Nextflow implementation) would be able to access them. I kicked the can down the road a bit thinking I'd just put this in and deal with the details of exactly where it is going to go once we've got everybody on board, but maybe I'll give it a bit more thought up-front to bolster the case for this approach.

**Comment by nedleitch** [comment]
> Code looks good, I think it will work just fine! I have 2 concerns though, that you probably thought about as well, so I'm curious to hear your thoughts:
>
Yeah had the same thoughts and decided to go with this approach because I felt that if it were easy enough there would be no reason not to use it, and then maybe we could just all get what we want. While that's an idealistic approach I don't think it is naively so because doing it this way saves so much work for everybody involved. I agree with your points but also thought it might be worth trying to build it the way we would like it to work first, then seeing if we can get some consensus rather than assuming we can't improve the process.

> 1. To make this project succeed, we should also include tools (and tool versions) that were released in the past (maybe a long time ago). How are you planning to deal with those, since you are working from the master branch of hmftools? With this approach it will be tough to "go back in history".

I figured we would want some of the old releases indeed. In my mind those would all be constructed from the existing Github releases. I had a fair bit of interaction with the Github APIs while I was writing this and am pretty confident I can put something together to scrape all the old releases, pluck out the JARs that were published and make Docker containers/publish Maven artifacts for them. 

> 2. Not everyone that releases tools in this repo will stick to "our" tag-release style. Also, there is no guarantee that everyone that releases tools on this repo will include the _same_ jar in the [releases page](https://github.com/hartwigmedical/hmftools/releases) that you get from running `mvn clean install` on the commit that was tagged. I believe there have been (please correct me if I'm wrong) multiple cases in the past where the commit that was tagged with the release does not compile or does not pass tests, and the jar that was included in the releases page was actually generated from a different commit, or from a code state that was never committed. Therefore to work this approach may require some behaviour changes that may or may not be feasible to ask from all contributors on this project.
> 

There have definitely been cases of this and it caused lots of headaches. The hours I spent generating md5s from buckets and images I will never get back! I think this is exactly why it's important to try to put down tooling that aligns with how we want to work rather than try to deal with exceptions that deviate from the norm. But this has been essentially my only hangup about doing it this way. I go back to what I said above which is I'd rather build something that aligns with how we want to work than try to accommodate a suboptimal de facto process.

> Because of points 1 and 2, I advocated an approach of putting the docker files in a separate repo, and adding some "subscription" logic, that creates a new docker image every time a new jar is added to the [releases page](https://github.com/hartwigmedical/hmftools/releases). Did you also consider this approach?

I didn't actually. Interesting idea. One thing about that is the releases are really quite mutable so it might be difficult to enforce a level of consistency that would probably be needed to get this right (artifact naming and uniqueness, release naming, etc).

File: hmftools-build.py
Comment (pauldwolfe) [reply hmftools-build.py:53]: I think we'll start with pushing internally and to DockerHub. DockerHub will be accessible to NextFlow. I guess the "external" image above is just a placeholder?

Bioconda, which is currently used by Stephen requires a separate "recipe" and approval process for each change, so let's consider it a non-starter for now.

**Comment by nedleitch** [comment]
I got a passing build that included the push to `docker.io`, so the image is now available on Dockerhub, but the repository is currently private. We will need to manually change the repository to public when we are ready.

But open question on this: the Docker container includes the JAR with all the dependencies, and it's not clear to me whether we might run afoul of legal issues if we publish this openly due to all the libraries we include.

Have been down this road in the past and I think it essentially comes down to having to verify everything has a permissive-enough licence. Any thoughts from either of you? Paul do you remember doing anything on this front before publishing any of the other public containers?

File: hmftools-build.py
Comment (pauldwolfe) [inline hmftools-build.py:44]: We'd want to uncomment this before merging?

File: hmftools-build.py
Comment (pauldwolfe) [inline hmftools-build.py:53]: private and public are terms we used to indicate one image has "private" stuff in it, while in this case its really the same image. I liked internal and external better.

File: hmftools-build.py
Comment (pauldwolfe) [inline hmftools-build.py:63]: Maybe GithubRelease so it's clear what this is doing?

**Review by pauldwolfe — APPROVED** [review]

**Comment by nedleitch** [comment]
> I got a passing build that included the push to `docker.io`, so the image is now available on Dockerhub, but the repository is currently private. We will need to manually change the repository to public when we are ready.
> 
> But open question on this: the Docker container includes the JAR with all the dependencies, and it's not clear to me whether we might run afoul of legal issues if we publish this openly due to all the libraries we include.
> 
> Have been down this road in the past and I think it essentially comes down to having to verify everything has a permissive-enough licence. Any thoughts from either of you? Paul do you remember doing anything on this front before publishing any of the other public containers?

I spoke with Paul about this and he confirmed these things had all been done before we started releasing the "with dependencies" JARs via Github releases so there should be nothing more to do here.

**Pull request merged**
